packages/panzer/disc-fe/src/Panzer_Workset_Builder_impl.hpp

```diff
@@ -222,55 +222,64 @@ panzer::buildBCWorkset(const WorksetNeeds & needs,
-  
+
   TEUCHOS_ASSERT(local_side_ids.size() == local_cell_ids.size());
   TEUCHOS_ASSERT(local_side_ids.size() == static_cast<std::size_t>(vertex_coordinates.extent(0)));
 
   // key is local face index, value is a pair of cell index and vector of element local ids
-  std::map<unsigned,std::vector<std::pair<std::size_t,std::size_t> > > element_list;
-  for (std::size_t cell=0; cell < local_cell_ids.size(); ++cell)
-    element_list[local_side_ids[cell]].push_back(std::make_pair(cell,local_cell_ids[cell])); 
+  std::map< unsigned, std::vector< std::pair< std::size_t, std::size_t>>> element_list;
+  for (std::size_t cell = 0; cell < local_cell_ids.size(); ++cell)
+    element_list[local_side_ids[cell]].push_back(std::make_pair(cell, local_cell_ids[cell]));
 
-  std::map<unsigned,panzer::Workset>& worksets = *worksets_ptr;
+  auto& worksets = *worksets_ptr;
 
-  // create worksets 
-  std::map<unsigned,std::vector<std::pair<std::size_t,std::size_t> > >::const_iterator side;
-  for (side = element_list.begin(); side != element_list.end(); ++side) {
+  // create worksets
+  for (const auto& side : element_list) {
 
-    std::vector<std::size_t>& cell_local_ids = worksets[side->first].cell_local_ids;
+    auto& cell_local_ids = worksets[side.first].cell_local_ids;
 
-    worksets[side->first].cell_vertex_coordinates = mdArrayFactory.buildStaticArray<double,Cell,NODE,Dim>("cvc",
-                                                          side->second.size(),
+    worksets[side.first].cell_vertex_coordinates = mdArrayFactory.buildStaticArray<double,Cell,NODE,Dim>("cvc",
+                                                          side.second.size(),
                                                           vertex_coordinates.extent(1),
                                                           vertex_coordinates.extent(2));
-    Workset::CellCoordArray coords = worksets[side->first].cell_vertex_coordinates;
+    auto coords_view = worksets[side.first].cell_vertex_coordinates;
 
-    for (std::size_t cell = 0; cell < side->second.size(); ++cell) {
-      cell_local_ids.push_back(side->second[cell].second);
+    for (std::size_t cell = 0; cell < side.second.size(); ++cell) {
+      cell_local_ids.push_back(side.second[cell].second);
 
-      for (std::size_t vertex = 0; vertex < Teuchos::as<std::size_t>(vertex_coordinates.extent(1)); ++ vertex)
-	for (std::size_t dim = 0; dim < Teuchos::as<std::size_t>(vertex_coordinates.extent(2)); ++ dim) {
-	  coords(cell,vertex,dim) = vertex_coordinates(side->second[cell].first,vertex,dim);
-        }
+      Kokkos::parallel_for(
+          vertex_coordinates.extent(1), KOKKOS_LAMBDA(std::size_t vertex)
+          {
+            const auto extent = Teuchos::as<std::size_t>(vertex_coordinates.extent(2));
+
+            for (std::size_t dim = 0; dim < extent; ++dim)
+            {
+              coords_view(cell, vertex, dim) = vertex_coordinates(side.second[cell].first, vertex, dim);
+            }
```

Comment:
Line 264 doesn't work for me and I'm amazed it did for you, but I can't compile.  @JacobDomagala what version of nvcc are you using?  

Anyway you can't have std::pair or std::vector so side.second[] has to be replaced with some view structure or the copy needs to happen on host.  I think the latter is better but,

Reply:
I'm using nvcc 11.4